src/Swift.Runtime/src/Library/ARC.cs

```diff
@@ -0,0 +1,168 @@
+// Licensed under the MIT License.
+
+using System;
+using System.Runtime.InteropServices;
+
+namespace Swift.Runtime;
+
+/// <summary>
+/// ARC is a class containing p/invokes for Swift Automatic Reference Counting and memory management.
+/// </summary>
+public static class Arc {
+    const string swiftCore = "/usr/lib/swift/libswiftCore.dylib";
+
+    /// <summary>
+    /// Retain a heap-allocated Swift object
+    /// </summary>
+    /// <param name="p">Pointer to an unmanaged Swift object, must be non-null.</param>
+    [DllImport(swiftCore)]
+    static extern void swift_retain(IntPtr p);
+
+    /// <summary>
+    /// Retain a heap-allocated Swift object
+    /// </summary>
+    /// <param name="p">Pointer to an unmanaged Swift object, must be non-null.</param>
+    /// <returns>The pointer passed in.</returns>
+    /// <exception cref="ArgumentOutOfRangeException">Throws if p is null</exception>
+    public static IntPtr Retain(IntPtr p)
+    {
+        if (p == IntPtr.Zero)
+            throw new ArgumentOutOfRangeException(nameof(p));
+        swift_retain(p);
+        return p;
+    }
+
+    /// <summary>
+    /// Check to see if a pointer is in the process of being deallocated.
+    /// </summary>
+    /// <param name="p">A non-null pointer to an unmanaged Swift object</param>
+    /// <returns>True if and only if the pointer in the process of being deallocated.</returns>
+    [DllImport(swiftCore)]
+    [return: MarshalAs(UnmanagedType.I1)]
+    static extern bool swift_isDeallocating(IntPtr p);
+
+    /// <summary>
+    /// Check to see if a pointer is in the process of being deallocated.
+    /// </summary>
+    /// <param name="p">A non-null pointer to an unmanaged Swift object</param>
+    /// <returns>True if and only if the pointer in the process of being deallocated.</returns>
+    /// <exception cref="ArgumentOutOfRangeException">Throws if p is null</exception>
+    public static bool IsDeallocating(IntPtr p)
+    {
+        if (p == IntPtr.Zero)
+            throw new ArgumentOutOfRangeException(nameof(p));
+        return swift_isDeallocating(p);
+    }
+
+    /// <summary>
+    /// Releases a heap-allocated Swift object
+    /// </summary>
+    /// <param name="p">Pointer to an unmanaged Swift object, must be non-null.</param>
+    [DllImport(swiftCore)]
+    static extern void swift_release(IntPtr p);
+
+    /// <summary>
+    /// Releases a heap-allocated Swift object
+    /// </summary>
+    /// <param name="p">Pointer to an unmanaged Swift object, must be non-null.</param>
+    /// <returns>The pointer passed in</returns>
+    /// <exception cref="ArgumentOutOfRangeException">Throws if p is null</exception>
+    /// <exception cref="Exception">Throws if p points to an object that has been deinitialized</exception>
+    public static IntPtr Release(IntPtr p)
+    {
+        if (p == IntPtr.Zero)
+            throw new ArgumentOutOfRangeException(nameof(p));
+        if (swift_isDeallocating(p))
```

Comment:
Why do we have to check this? I would expect Swift to do this - is it because we don't have a way to catch that error?

Reply:
Swift does not check for null in some cases. In the cases that it checks, there is no error reported. In the cases that it doesn't check, the code will crash hard if we're lucky. In the case of checking for isDeallocating, this is something that doesn't happen in Swift by contract because the compiler generates the calls to that code. If this gets triggered, we've done something horribly wrong.
Should I put this in a `#if DEBUG`?

Reply:
I think it's OK to keep it. The perf difference it probably unobservable, and if it does hit, we will at least have some more meaningful info about the failure.